File: xcpp/XRemoteInput.cpp
#include "pch.h"
#include "XRemoteInput.h"
#include "XLog.h"

using namespace xcpp;

XRemoteInput::XRemoteInput(XInput& xinput, XOutput* xoutput) : XInput(xinput)
{
	mInput = &xinput;
	mOutput = xoutput;
}

XRemoteInput::~XRemoteInput()
{
	delete mOutput;
}

XRemoteInput::XRemoteInput(const XRemoteInput& other) : XInput(other)
{
	mInput = other.mInput;
	mOutput = other.mOutput;
}

bool XRemoteInput::xread()
{
	XLog xlog(L"XRemoteInput::xread");
	bool bRes = mInput->xread();
	if (bRes)
	{
		mOutput->xfirstString(mInput->xfirstString());
		map<wstring, wstring> data;
		mInput->xreadStrings(data);
		mOutput->xwriteStrings(data);
		bRes = mOutput->xwrite();
	}
	return bRes;
}

bool XRemoteInput::xclose()
{
	XLog xlog(L"XRemoteInput::xclose");
	bool bRes = mOutput->xclose();
	if (bRes)
	{
		mInput->xclose();
	}
	return bRes;
}

File: XCS/XRemoteInput.cs
using System.Collections.Generic;

namespace XCS
{
    public class XRemoteInput : XInput
    {
        public XRemoteInput(XInput xinput, XOutput xoutput) : base()
        {
            XInput = xinput;
            XOutput = xoutput;
        }
        public XInput XInput { get; protected set; }
        public XOutput XOutput { get; protected set; }

        public override bool XClose()
        {
            return true;
        }

        public override bool XReadStrings(out List<string> strings)
        {
            bool bRes = XInput.XReadStrings(out strings);
            if(bRes)
            {
                bRes = XOutput.XWriteStrings(strings);
            }
            return bRes;
        }
    }
}
